update FastTimerService for CMSSW 9.1.x

fwyzard opened the pull request:
EDMModuleType: enum and functions to classify edm module types

ProcessCallGraph: helper class to manage the process description and modules call graph

FastTimerService: rewrite to support multiple subprocesses, multiple threads per stream and unscheduled execution    
  - defer module dependencies, path structure and process id to the `ProcessCallGraph` class
  - add stubs for all relevant signals
  - remove plots vs luminosity, and the plugins that were used to book and fill such plots; instead, one should use the `FastTimerServiceClient` providing similar functionality in the harvesting step
  - remove the obsolete `FastTimer` classes
  - remove information for overhead and by "module type", and the details that do not make sense with multiple threads per stream
  - reimplement DQM plots and query methods

fwyzard committed:
ProcessCallGraph: helper class to manage the process description and modules call graph

fwyzard committed:
FastTimerService: rewrite to support multiple subprocesses, multiple threads per stream and unscheduled execution

  - defer module dependencies, path structure and process id to ProcessCallGraph;
  - add stubs for all relevant signals;
  - remove plots vs luminosity, and the plugins that were used to book and fill such plots;
    instead, one should use the FastTimerServiceClient providing similar functionality in the
    harvesting step;
  - remove query methods, overhead information, information by "module type", and the details
    that do not make sense with with multiple threads per stream.

fwyzard committed:
EDMModuleType: enum and functions to classify edm module types

fwyzard committed:
remove obsolete FastTimer class

fwyzard committed:
FastTimerService: reimplement DQM plots

  - book all plots during the main process preStreamBeginRun
  - fill all plots during the last process postEvent

fwyzard committed:
FastTimerService: reimplement query methods

fwyzard committed:
FastTimerService: reimplement summaries and reorganise the DQM folder structure

fwyzard committed:
FastTimerService: fix merging plots when subprocesses are present

fwyzard committed:
FastTimerService: fix range of module counters

A TH1F has 7 significant digits, while a 24-hour long run could process
order of 10 billion events; a 64-bit long integer would work and might
be better suited than a double, but there is no "TH1L" in ROOT.

fwyzard committed:
FastTimerService: monitor a configurable set of modules

fwyzard committed:
FastTimerService: add debugging for thread ids

fwyzard committed:
FastTimerService: fix concurrency issue

fwyzard commented:
please test

fwyzard commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @muzaffar, @davidlange6, @smuzaffar

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

HLTrigger/Timer


@cmsbuild, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@muzaffar, @davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

cmsbuild commented:
-1

Tested at: cfce07fbfb9b256b40713719b83f059c9be07767

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19270/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
4.22 step2
<pre>runTheMatrix-results/4.22_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC/step2_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC.log

</pre>140.53 step2
<pre>runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/step2_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI.log

</pre>1000.0 step2
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step2_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>1001.0 step2
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/step2_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.log

</pre>1003.0 step2
<pre>runTheMatrix-results/1003.0_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM/step2_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM.log

</pre>4.53 step3
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT/step3_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT.log

</pre>8.0 step3
<pre>runTheMatrix-results/8.0_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS/step3_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS.log

</pre>1306.0 step3
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step3_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>10042.0 step3
<pre>runTheMatrix-results/10042.0_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step3_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>20034.0 step3
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D7_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D7+RecoFullGlobal_2023D7+HARVESTFullGlobal_2023D7/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D7_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D7+RecoFullGlobal_2023D7+HARVESTFullGlobal_2023D7.log

</pre>136.731 step3
<pre>runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step3_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

</pre>9.0 step3
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step3_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>10024.0 step3
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>50202.0 step3
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step3_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

</pre>10824.0 step3
<pre>runTheMatrix-results/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018.log

</pre>25.0 step3
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>23234.0 step3
<pre>runTheMatrix-results/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D8_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D8+RecoFullGlobal_2023D8+HARVESTFullGlobal_2023D8/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D8_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D8+RecoFullGlobal_2023D8+HARVESTFullGlobal_2023D8.log

</pre>21234.0 step3
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D4_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D4+RecoFullGlobal_2023D4+HARVESTFullGlobal_2023D4/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D4_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D4+RecoFullGlobal_2023D4+HARVESTFullGlobal_2023D4.log

</pre>25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake2.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:43:35 2017-date Wed Apr 19 19:40:03 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:44:02 2017-date Wed Apr 19 19:40:08 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:46:18 2017-date Wed Apr 19 19:40:13 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:48:01 2017-date Wed Apr 19 19:40:18 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:48:04 2017-date Wed Apr 19 19:40:25 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun2016.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:49:10 2017-date Wed Apr 19 19:40:33 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:44:43 2017-date Wed Apr 19 19:40:34 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:47:12 2017-date Wed Apr 19 19:40:38 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:47:19 2017-date Wed Apr 19 19:40:43 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake2.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:47:53 2017-date Wed Apr 19 19:40:44 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake1.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:47:35 2017-date Wed Apr 19 19:40:49 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:43:47 2017-date Wed Apr 19 19:40:55 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:45:54 2017-date Wed Apr 19 19:40:59 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake1.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:44:00 2017-date Wed Apr 19 19:41:03 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:48:21 2017-date Wed Apr 19 19:41:05 2017 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun2016.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 19:47:26 2017-date Wed Apr 19 19:41:07 2017 s - exit: 18688

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

fwyzard committed:
FastTimerService: ignore obsolete parameters and fix test file

fwyzard committed:
FastTimerService: fix user configurations

cmsbuild commented:
Pull request #18411 was updated. @perrotta, @cmsbuild, @ghellwig, @silviodonato, @arunhep, @cerminar, @dmitrijus, @Martin-Grunewald, @franzoni, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
+!

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 9b5dc989893741089533f58a3e28968f72cab0c2

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19276/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
4.53 step3
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT/step3_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT.log

</pre>1000.0 step2
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step2_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>1001.0 step2
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/step2_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.log

</pre>1003.0 step2
<pre>runTheMatrix-results/1003.0_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM/step2_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM.log

</pre>1306.0 step3
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step3_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>10042.0 step3
<pre>runTheMatrix-results/10042.0_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step3_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>136.731 step3
<pre>runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step3_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

</pre>20034.0 step3
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D7_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D7+RecoFullGlobal_2023D7+HARVESTFullGlobal_2023D7/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D7_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D7+RecoFullGlobal_2023D7+HARVESTFullGlobal_2023D7.log

</pre>9.0 step3
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step3_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>50202.0 step3
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step3_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

</pre>25.0 step3
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>10024.0 step3
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>10824.0 step3
<pre>runTheMatrix-results/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018.log

</pre>23234.0 step3
<pre>runTheMatrix-results/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D8_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D8+RecoFullGlobal_2023D8+HARVESTFullGlobal_2023D8/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D8_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D8+RecoFullGlobal_2023D8+HARVESTFullGlobal_2023D8.log

</pre>21234.0 step3
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D4_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D4+RecoFullGlobal_2023D4+HARVESTFullGlobal_2023D4/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D4_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D4+RecoFullGlobal_2023D4+HARVESTFullGlobal_2023D4.log

</pre>25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake2.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:19:51 2017-date Wed Apr 19 23:13:03 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:19:19 2017-date Wed Apr 19 23:13:04 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:17 2017-date Wed Apr 19 23:13:07 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:44 2017-date Wed Apr 19 23:13:08 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:42 2017-date Wed Apr 19 23:13:17 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun2016.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:22:02 2017-date Wed Apr 19 23:13:21 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:19:18 2017-date Wed Apr 19 23:13:22 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:04 2017-date Wed Apr 19 23:13:26 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:14 2017-date Wed Apr 19 23:13:30 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake2.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:21 2017-date Wed Apr 19 23:13:35 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake1.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:20 2017-date Wed Apr 19 23:13:38 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:18:05 2017-date Wed Apr 19 23:13:46 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:23 2017-date Wed Apr 19 23:13:51 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake1.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:19:48 2017-date Wed Apr 19 23:13:54 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:21:47 2017-date Wed Apr 19 23:13:59 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun2016.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Wed Apr 19 23:22:33 2017-date Wed Apr 19 23:14:02 2017 s - exit: 18688

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

dmitrijus commented:
+1

Test error is HLT specific.

cmsbuild commented:
Pull request #18411 was updated. @perrotta, @cmsbuild, @ghellwig, @silviodonato, @arunhep, @cerminar, @dmitrijus, @Martin-Grunewald, @franzoni, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

fwyzard commented:
ok, now I got `runTheMatrix.py -l limited -j4` to run successfully locally ...

fwyzard commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

dmitrijus commented:
+1

cmsbuild commented:
-1

Tested at: 1dbbbf1e7b2bea31d9a0ab2047d21d346fad83d6

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19285/summary.html

I found follow errors while testing this PR

**Failed tests:  AddOn**

* **AddOn**:

I found errors in the following addon tests:

cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_Fake2.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:41:21 2017-date Thu Apr 20 12:37:45 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:41:37 2017-date Thu Apr 20 12:37:48 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:43:50 2017-date Thu Apr 20 12:37:51 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:45:38 2017-date Thu Apr 20 12:37:52 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:45:45 2017-date Thu Apr 20 12:37:59 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun2016.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:46:32 2017-date Thu Apr 20 12:38:06 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:41:54 2017-date Thu Apr 20 12:38:08 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_Fake.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:44:44 2017-date Thu Apr 20 12:38:11 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:44:37 2017-date Thu Apr 20 12:38:15 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_Fake2.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:45:28 2017-date Thu Apr 20 12:38:20 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_Fake1.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:45:20 2017-date Thu Apr 20 12:38:22 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_Fake.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:41:17 2017-date Thu Apr 20 12:38:29 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:43:17 2017-date Thu Apr 20 12:38:34 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_Fake1.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:41:29 2017-date Thu Apr 20 12:38:39 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:45:26 2017-date Thu Apr 20 12:38:41 2017 s - exit: 18688
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-19-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun2016.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Apr 20 12:45:20 2017-date Thu Apr 20 12:38:45 2017 s - exit: 18688

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Pull request #18411 was updated. @perrotta, @cmsbuild, @ghellwig, @silviodonato, @arunhep, @cerminar, @dmitrijus, @Martin-Grunewald, @franzoni, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

fwyzard commented:
...and this should fix also the addons

fwyzard commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fwyzard commented:
@Martin-Grunewald @silviodonato does it (finally) look good to you ?

cmsbuild commented:
+1
Tested at: 62e51ac8983934ae5262d8a5fe82138d1dc859a1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19292/summary.html

cmsbuild commented:
Comparison job queued.

fwyzard committed:
FastTimerService: extend highlighting to multiple groups of modules

fwyzard committed:
FastTimerService: make some plots optional

  - plots for each individual module are enabled by
      FastTimerService.enableDQMbyModule = cms.untracked.bool(True)

  - plots vs lumisection are enabled by
      FastTimerService.enableDQMbyLumiSection = cms.untracked.bool(True)

cmsbuild commented:
Pull request #18411 was updated. @perrotta, @cmsbuild, @ghellwig, @silviodonato, @arunhep, @cerminar, @dmitrijus, @Martin-Grunewald, @franzoni, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
+1

fwyzard commented:
If no bugs appear, I think I should be done for now.
I think all the features we needed more or less urgently are now in. Further developments will happen in a different branch.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 489726e4972055fdefee7ca26f1867eecfa26a71
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19302/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19302/summary.html

Comparison Summary:
 - You potentially added 207 lines to the logs
 - Reco comparison results: 1669 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1820537
 - DQMHistoTests: Total failures: 20966
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1799398
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

dmitrijus commented:
+1

perrotta commented:
+1
Updates to the FastTimerService

davidlange6 commented:
assign core
(looks good to my untrained eye but since a different PR seems related I wanted to connect the loop)

cmsbuild commented:
New categories assigned: core

@Dr15Jones,@smuzaffar you have been requested to review this Pull request/Issue and eventually sign? Thanks

fwyzard commented on HLTrigger/Timer/src/FastTimerService.cc:
```diff
@@ -38,1425 +39,1365 @@
-    for (auto & keyval: m_stream.moduletypes) {
-      std::string const & label  = keyval.first;
-      ModuleInfo  const & module = keyval.second;
-      out << "FastReport              " << std::right << std::setw(10) << module.summary_active  / (dou
```
@Dr15Jones turns out that this is not good enough :-(

In a job configured to run with N threads, I have sometimes seen an (N+1)th or (N+2)th thread appear;  `threadId()` nicely assigns them new ids, but then `threads_` is not big enough.

Here in an excerpt with some debugging messages added in:
```
%MSG-i ThreadSetup:  (NoModuleName) 22-Apr-2017 01:11:27 CEST pre-events
setting # threads 8
%MSG
%MSG-i StreamSetup:  (NoModuleName) 22-Apr-2017 01:11:27 CEST pre-events
setting # streams 8
%MSG
...
thread 139795564551296 assigned unique id 0
thread 139795564551296 using unique id 0 / 7
thread 139795564551296 using unique id 0 / 7
...
thread 139793846748928 assigned unique id 1
thread 139793846748928 using unique id 1 / 7
thread 139793899976448 assigned unique id 2
thread 139793899976448 using unique id 2 / 7
thread 139793899976448 using unique id 2 / 7
thread 139793846748928 using unique id 1 / 7
thread 139793846748928 using unique id 1 / 7
thread 139793867728640 assigned unique id 4
thread 139793650677504 assigned unique id 5
thread 139793889486592 assigned unique id 3
thread 139793640187648 assigned unique id 7
thread 139793857238784 assigned unique id 6
thread 139793650677504 using unique id 5 / 7
thread 139793867728640 using unique id 4 / 7
...
```
and then much later
```
...
thread 139793878218496 assigned unique id 8
thread 139793878218496 using unique id 8 / 7
thread 139793835489024 assigned unique id 9
thread 139793889486592 using unique id 3 / 7
thread 139793889486592 using unique id 3 / 7
thread 139793889486592 using unique id 3 / 7
thread 139793889486592 using unique id 3 / 7
thread 139793889486592 using unique id 3 / 7
thread 139793835489024 using unique id 9 / 7
...
```
soon after leading to
```
----- Begin Fatal Exception 22-Apr-2017 01:13:10 CEST-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 165121 lumi: 62 event: 23684942 stream: 4
   [1] Running path 'pathALCARECODtCalib'
   [2] Prefetching for module GoodVertexFilter/'primaryVertexFilter'
   [3] Prefetching for module RecoChargedRefCandidatePrimaryVertexSorter/'offlinePrimaryVertices'
   [4] Prefetching for module FastjetJetProducer/'ak4CaloJetsForTrk'
   [5] Prefetching for module CaloTowersCreator/'caloTowerForTrk'
   [6] Prefetching for module EcalRecHitProducer/'ecalRecHit'
   [7] Prefetching for module EcalUncalibRecHitProducer/'ecalMultiFitUncalibRecHit'
   [8] Calling method for module EcalRawToDigi/'ecalDigis'
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 9) >= this->size() (which is 8)
----- End Fatal Exception -------------------------------------------------
```

Do you have any ideas what might be going on ?
So far I have thought of
  - TBB is removing old threads and adding new threads to the pool (possibility suggested by the existence of a `tbb:task_scheduler_observer` class, see https://software.intel.com/en-us/node/506314);
  - some of the signals end up on threads that are handling xrootd or other I/O (I don't think I have see this happen with an `EmptysSurce`).

Either way, the workarounds I could think of are
  - use a `thread local Measurement` instead of a `std::vector<Measurement>`: this will give static objects which would not be suited for a normal module, but since a Service is kind of a singleton anyway, it could work;
  - use a `tbb::concurrent_hash_map<thread::id,Measurement>` keyed on `this_thread::get_id()`: this should work, no idea on its efficiency;
  - use a `tbb::enumerable_thread_specific<Measurement>`: this should do the same as the previous one, but is specialised for this use case.
 
Comments or suggestions ?

fwyzard replied:
`tbb::enumerable_thread_specific<Measurement>` works...

fwyzard replied:
OK, it does look like threads are being swapped in and out.
I've changed the code to use `tbb::enumerable_thread_specific<Measurement>`, so it can support an arbitrary number of threads, and reproduced the logs.

There are the line ranges of the log file where each thread is used:
```
0th thread 140223481818240 used in range    513 - 207612
1st thread 140221795268352 used in range    525 - 207588
2nd thread 140221806274304 used in range    528 -  13908
3rd thread 140221784778496 used in range    533 - 207540
4th thread 140221773510400 used in range    536 - 207604
5th thread 140221752530688 used in range    551 -  13926
6th thread 140221816764160 used in range    565 - 207558
7th thread 140221763020544 used in range    649 - 207594
8th thread 140216454346496 used in range  13927 - 207568
9th thread 140216443856640 used in range  14063 - 207532

```

It looks like the 2nd and 5th threads were dropped at some point, and replaced by new threads.

fwyzard replied:
Anyway, the fix is here: a569e589c4641f276678fb31ab0952658d800ad3 .

@davidlange6, should I update this PR or make a new one (so it won't need all other signatures after this one is merged) ?

Dr15Jones replied:
Is this really ignored?

Dr15Jones replied:
The framework was changed so all child SubProcesses are launched simultaneously. Therefore the SubProcess with the largest value is not guaranteed to be the last one to finish. 
I could make it such that the parent processes doesn't send its 'postStreamEndRun' signal until all its child SubProcesses have finished.

Dr15Jones replied:
I'm pretty sure the change is only in the CMSSW_9_1 release.

Dr15Jones replied:
This also has the same problem about the concept of 'last'

Dr15Jones replied:
Is this ignored?

Dr15Jones replied:
Same problem with meaning of 'last'

Dr15Jones replied:
Is this ignored?

Dr15Jones replied:
I think this is a race condition. Since mulitple SubProcesses could reach `postEvent` at the same time and then try to update `stream.total` simultaneously.

Dr15Jones replied:
same problem with `last`.

Dr15Jones replied:
Another idea could be to use a `tbb::concurrent_vector` and extend it when needed.

fwyzard committed:
FastTimerService: fix concurrency (again)

Use a tbb::enumerable_thread_specific for thread-local storage.
This gives one lazily-allocated object per thread.

fwyzard committed:
FastTimerService: move implementation code to the .cc file

Dr15Jones commented on HLTrigger/Timer/interface/FastTimerService.h:
```diff
@@ -86,690 +82,488 @@ class FastTimerService {
-  std::vector<TimingPerProcess>                 m_job_summary_perprocess;       // per-process time accounting per-job
-  std::mutex                                    m_summary_mutex;                // synchronise access to the summary objects across different threads
+  ResourcesPerJob               job_summary_;                   // whole event time accounting per-job
+  std::vector<ResourcesPerJob>  run_summary_;                   // whole event time accounting per-run
```
Does this have one entry per run seen in the job?

fwyzard replied:
No, this has as many elements as `edm::service::SystemBounds::maxNumberOfConcurrentRuns()`, and is indexed via `edm::GlobalContext::runIndex()` or `edm::StreamContext::runIndex()` .

My assumption is that there are never more than `maxNumberOfConcurrentRuns()` "in flight" at any given moment, and that any access happens between `preGlobalBeginRun` and `postGlobalEndRun`.

Dr15Jones replied:
Your assumption is correct.

Dr15Jones commented on HLTrigger/Timer/interface/FastTimerService.h:
```diff
@@ -86,690 +82,488 @@ class FastTimerService {
-  //  - all endpaths  (available after all endpaths have run)
-  //  - processing    (available after the event has been processed)
+public:
+  // query the time spent in a module/path/process (available after it has run)
```
I'm concerned that the `query` calls may be causing undefined behavior since you have no explicit synchronization going on between the threads that are writing the data and those calling `query`. TBB tasks internally do a memory barrier (which is a synchronization) before the task starts and again after the task finishes. This guarantees that if a task depends on the results of another task which finished before it that the memory is all synchronized. As far as I can tell, your `query` calls have no such guarantees.

fwyzard replied:
So, based on your comment and assuming that the query is for a module in the same stream
  - if the module has already run, it was on a different task, so they should be safe because of TBB synchronisation;
 - if the module is running or has not yet run, query should return 0, because the value is store only after the end of the module.

Anyway, I am not very fond of the query functions myself - eventually I would like to make the information for all Paths available in the event (in a similar way that the TriggerResults are made available).

Dr15Jones commented on HLTrigger/Timer/src/FastTimerService.cc:
```diff
@@ -38,1425 +39,1365 @@
-    const std::string label = "the whole job";
-    for (unsigned int pid = 0; pid < m_process.size(); ++pid)
-      printProcessSummary(m_job_summary, m_job_summary_perprocess[pid], label, m_process[pid].name);
+  ignoredSignal(__func__);
```
Is this really ignored since you have code later in this function?

fwyzard replied:
Sorry, `ignoredSignal()` and `unsupportedSignal()` are more of a work-in-progress thing.

Eventually, I would like to assign all the time spent in the job to "something" - be it a run or lumi transition, and event, the framework itself, or account it as idle or blocked threads (though I don't really know how to do the last part).

For the moment, `unsupportSignal` are the signals I don't know what they do, while `ignoredSignal` are the signals I think I know what they do, but I am not handling yet.

Dr15Jones commented:
Rather than connect signals and then call `ignoredSignal` it would be better to disconnect the signals from the `ActivityRegistry` since you are forcing the framework to call each of those unnecessary callbacks.

fwyzard commented on HLTrigger/Timer/interface/FastTimerService.h:
```diff
@@ -86,690 +82,488 @@ class FastTimerService {
+  void preModuleStreamEndLumi(edm::StreamContext const&, edm::ModuleCallingContext const&);
+  void postModuleStreamEndLumi(edm::StreamContext const&, edm::ModuleCallingContext const&);
+
+  void preModuleEventPrefetching(edm::StreamContext const&, edm::ModuleCallingContext const&);
```
OK, thanks. Actually, what do these signal *mean* ?

Dr15Jones replied:
Before running a module, the framework will prefetch all data the module says it will consume. The `preModuleEventPrefetching` happens when the framework has decided the module should be run and it is going to start prefetching its data. `postModuleEventPrefetching` is signaled when all the prefetching has completed and the framework can now schedule to run the module. For global and stream module, the framework will start them immediately. For legacy and one modules, the framework will wait until it can obtain the shared resource.

fwyzard replied:
Thanks.

fwyzard replied:
Can I ask also about `pre`/`postEventReadFromSource` ?

Dr15Jones replied:
`pre/postEventReadFromSource` happens when a module asks for data and the framework starts a 'delayed read' of the data from the source. E.g., someone dereferences a `edm::Ref` or just has a `consumes` and we are prefetching the data. Those signals happens right before asking the source to read the data and right after it has finished reading. These are called only after the task has acquired the 'shared resource' for the source.

fwyzard replied:
Thanks.

fwyzard commented:
I see... OK, I'll replace most ignored/unsupported signals with a comment or a macro, as a reminder.

fwyzard commented on HLTrigger/Timer/interface/FastTimerService.h:
```diff
@@ -28,23 +34,13 @@
 
 
 /*
 procesing time is diveded into
```
Sure.

fwyzard commented:
Hi @Dr15Jones , I'm not sure why, but I see many of your comments attached to the wrong part of the code, so I don't really know what some of them are about.

If I understand correctly, the problem is that the Subprocesses can run in parallel, so
  - two of them could complete at the same time, creating a race condition in the update of the total event statistics
  - there is no simple way to know when all of them are done

For the first issue I guess I can add a mutex around the update - the conflicts should be rare enough and the operation fast enough.

For the more general problem of knowing when all of them are done, either the framework can add a signal for it (if there is a point where that is known to be the case), or I could use something like a vector<atomic_bool> to keep track of the status of all subprocesses, and do the accounting only after all are done.

fwyzard commented:
> For the more general problem of knowing when all of them are done, either the framework can add a signal for it (if there is a point where that is known to be the case), or I could use something like a `vector<atomic_bool>` to keep track of the status of all subprocesses, and do the accounting only after all are done.

OK, that's overly complicated and won;t actually work. Just use an `atomic<unsigned int>` to keep track of how many subprocesses are complete, and trigger the accounting only in the thread that is responsible for completing the last one.

Dr15Jones commented:
Counting the number of processes which have completed a given transition does seem like the best choice.

fwyzard commented:
@Dr15Jones , how do I know when I can reset those counters ?
  - for the (per-stream) counter used in `postEvent`, I think I can reset it at `preSourceEvent`

  - for the (per-stream) counter used in `postStreamEndLumi` and `postStreamEndRun`, I cannot use a Source transition, because those seem to be global, not per-stream

  - for the (per-run) counter used in `postGlobalEndRun`, I thought I could reset it at `preSourceRun` - however that doesn't work because the signal does not have a global context or other way to extract the "run index"

Does the framework guarantee that the `preStreamBeginLumi`, `preStreamBeginRun`, `preGlobalBeginRun` signals are called for the main process first ?
Or, that they are called for all processes before any of the corresponding `postStreamEndLumi` etc are called ?

fwyzard committed:
FastTimerService: make check-for-last-subprocess more robust

fwyzard committed:
FastTimerService: do not watch ignored signals

cmsbuild commented:
Pull request #18411 was updated. @perrotta, @cmsbuild, @ghellwig, @silviodonato, @arunhep, @cerminar, @dmitrijus, @Martin-Grunewald, @franzoni, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

fwyzard commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 611b2325596795ceb3bc0b857e6cf3b5c34efd6c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19337/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19337/summary.html

Comparison Summary:
 - You potentially added 6 lines to the logs
 - Reco comparison results: 1643 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1774322
 - DQMHistoTests: Total failures: 20880
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1753269
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

Dr15Jones commented:
> Does the framework guarantee that the preStreamBeginLumi, preStreamBeginRun, preGlobalBeginRun signals are called for the main process first ?
Or, that they are called for all processes before any of the corresponding postStreamEndLumi etc are called ?

The way it is setup now (and the way I believe we will keep it) is the 'parent' process has to complete its transition before starting its 'child' process transition. I.e. the top level process will do 'pre/post' for (global|stream)(Begin|End)(Run|LuminosityBlock) before it will simultaneously start the corresponding transitions for all its SubProcesses and they then do their 'pre/post'.

fwyzard commented:
> The way it is setup now (and the way I believe we will keep it) is the 'parent' process has to complete its transition before starting its 'child' process transition.

OK, thanks. Then what I'm doing [here](https://github.com/cms-sw/cmssw/pull/18411/files#diff-1f221fefd6829c1e8a4d9b15bf1369b1R1304) should be OK.

fwyzard commented:
Bonus questions...
  - are the `pre`/`postSourceEvent` transitions per stream, or global ?
  - are the `pre`/`postSourceRun` and `pre`/`postSourceLumi` transitions per stream, or global ?
do they signal around the begin run/lumi or around the whole run/lumi ?
that is, do they mean `pre`/`postSourceBeginRun`, or `preSourceBeginRun`/`postSourceEndRun` ?

Dr15Jones commented:
> are the pre/postSourceEvent transitions per stream, or global ?

Any signal that takes a `StreamID` as an argument is per stream. In the case of those exact signals, the source has 'peeked' ahead and identified that the next record to process is an event and now the framework has assigned that event to the given stream and the source is now filling in the needed information into the Event class.

>are the pre/postSourceRun and pre/postSourceLumi transitions per stream, or global ?

They are global. The source has 'peeked' ahead and seen that the next transition is a Run or Lumi and now the framework has told the Source to read the necessary information and fill in the Run or Lumi object.

>do they signal around the begin run/lumi or around the whole run/lumi ?
that is, do they mean pre/postSourceBeginRun, or preSourceBeginRun/postSourceEndRun ?

They signal around just the call to the Source, so `pre/postSourceBeginRun`. To the source, there is no concept of 'end' of a transition. Once it has filled a record it never looks at it again.

fwyzard commented:
> They are global. The source has 'peeked' ahead and seen that the next transition is a Run or Lumi and now the framework has told the Source to read the necessary information and fill in the Run or Lumi object.

Do you think it would be possible to add a `GlobalContext` to those signals ?

fwyzard commented:
@Dr15Jones , with the latest changes, do you think the PR can go in ?
I know I may not have addressed the query* methods satisfactorily; they are not widely used, and I plan to address them somehow at a later time.

fwyzard commented:
+1

Dr15Jones commented:
Only the outer most process has a 'source' so by construction all source based signals come from that process. As such, I'm not sure what you're trying to accomplish by changing the signature of (pre|post)Source(Run|Lumi)?

fwyzard commented:
I would like to know the run number and/or runIndex that is about to start.

Dr15Jones commented:
At preSourceRun (and similary for preSourceLumi) the framework only knows the type of the next transition, not what run (or lumi) number which it will have. The framework should know which index to which it will be assigned at that time.

Dr15Jones commented on HLTrigger/Timer/src/FastTimerService.cc:
```diff
@@ -38,1425 +36,1603 @@
+///////////////////////////////////////////////////////////////////////////////
 
+// convert any std::chrono::duration to milliseconds
+template <class Rep, class Period>
```
How about putting these two functions in an anonymous namespace? Otherwise they go into the global namespace and could potentially clash with a function in another shared library.

Dr15Jones commented:
Visual inspection only turned up the one issue with `ms` being in the global namespace. Once that is changed, I will give my approval.

fwyzard committed:
move global functions to an anonymous namespace

cmsbuild commented:
Pull request #18411 was updated. @perrotta, @cmsbuild, @ghellwig, @silviodonato, @arunhep, @cerminar, @dmitrijus, @Martin-Grunewald, @franzoni, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

fwyzard commented:
@Dr15Jones done.
@davidlange6 do you mind bypassing the other signatures? there were no changes outside of the HLTrigger/Timer package since the PR was signed by Reco and DQM.

Dr15Jones commented:
+1

cmsbuild commented:
-1

Tested at: de21c55c6d8d0cc98057c3aa4c5644a44e53d9a3

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19354/summary.html

I found follow errors while testing this PR

**Failed tests:  AddOn**

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_PRef --relval 9000,50 --datatier "RAW" --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2017 --magField 38T_PostLS1 --eventcontent RAW --fileout file:RelVal_Raw_PRef_DATA.root --filein /store/data/Run2016B/JetHT/RAW/v1/000/272/762/00000/C666CDE2-E013-E611-B15A-02163E011DBE.root : FAILED - time: date Tue Apr 25 00:07:21 2017-date Mon Apr 24 23:58:10 2017 s - exit: 23552
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-24-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Tue Apr 25 00:07:21 2017-date Mon Apr 24 23:58:10 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:PRef,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_PRef --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_PRef_DATA.root --fileout file:RelVal_Raw_PRef_DATA_HLT_RECO.root : FAILED - time: date Tue Apr 25 00:07:21 2017-date Mon Apr 24 23:58:10 2017 s - exit: 21504
cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_PIon --relval 9000,50 --datatier "RAW" --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2017 --magField 38T_PostLS1 --eventcontent RAW --fileout file:RelVal_Raw_PIon_DATA.root --filein /store/data/Run2016B/JetHT/RAW/v1/000/272/762/00000/C666CDE2-E013-E611-B15A-02163E011DBE.root : FAILED - time: date Tue Apr 25 00:11:53 2017-date Tue Apr 25 00:02:42 2017 s - exit: 23552
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-24-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Tue Apr 25 00:11:53 2017-date Tue Apr 25 00:02:42 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2017 --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_PIon_DATA.root --fileout file:RelVal_Raw_PIon_DATA_HLT_RECO.root : FAILED - time: date Tue Apr 25 00:11:53 2017-date Tue Apr 25 00:02:42 2017 s - exit: 21504
cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_1_X_2017-04-24-1100/src/Utilities/ReleaseScripts/scripts/read312RV_cfg.py : FAILED - time: date Tue Apr 25 00:14:40 2017-date Tue Apr 25 00:07:26 2017 s - exit: 23552
cmsDriver.py RelVal -s L1REPACK:GT1   --data --scenario=pp -n 10 --conditions auto:run1_hlt_Fake --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --fileout file:RelVal_Raw_Fake_DATA.root --filein /store/data/Run2012A/MuEG/RAW/v1/000/191/718/14932935-E289-E111-830C-5404A6388697.root : FAILED - time: date Tue Apr 25 00:19:08 2017-date Tue Apr 25 00:10:51 2017 s - exit: 23552
cmsRun /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_1_X_2017-04-24-1100/src/HLTrigger/Configuration/test/OnLine_HLT_Fake.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Tue Apr 25 00:19:08 2017-date Tue Apr 25 00:10:51 2017 s - exit: 21504
cmsDriver.py RelVal -s HLT:Fake,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run1_data_Fake --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --processName=HLTRECO --filein file:RelVal_Raw_Fake_DATA.root --fileout file:RelVal_Raw_Fake_DATA_HLT_RECO.root : FAILED - time: date Tue Apr 25 00:19:08 2017-date Tue Apr 25 00:10:51 2017 s - exit: 21504

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19354/summary.html

Comparison Summary:
 - You potentially added 73 lines to the logs
 - Reco comparison results: 69 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1774356
 - DQMHistoTests: Total failures: 29408
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1744775
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

ghellwig commented:
+1

dmitrijus commented:
+1

perrotta commented:
+1
Updates to the FastTimerService

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (but tests are reportedly failing). This pull request requires discussion in the ORP meeting before it's merged. @muzaffar, @davidlange6, @smuzaffar

silviodonato commented:
addOn tests failed because of a EOS glitch
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18411/19354/addOnTests/hlt_data_PRef/cmsDriver.py_RelVal_-s_L1REPACK:Full_--data_--scenario=pp_-n_10_--conditions_auto:run2_hlt_PRef_--relval_9000,50_--datatier_RAW_--customise=HLTrigge.log
